### ISSUE TITLE
question about Quantize BERT?

### ISSUE THREAD

**Issue by RyanHuangNLP** [NervanaSystems/nlp-architect#86]
I find that the excited codes about [Quantize BERT](http://nlp_architect.nervanasys.com/quantized_bert.html), use int 8, but only loss a little accuracy, but after I train the mrpc with the example code, I find that the checkpoint file is 418mb, is it normal? I think it will be two times smaller.

is it the int 8 method describe similar with [Vector Neural Network Instructions Enable Int8 AI Inference on Intel Architecture](https://www.intel.ai/vnni-enables-inference/)

And there is any example about [Knowledge Distillation using Transformers](http://nlp_architect.nervanasys.com/transformers_distillation.html)?

**Comment by ofirzaf** [comment]
Hi,

When training quantized BERT the weights are saved in fp32. Therefore, there is no change in the checkpoint size, it might even be slightly bigger since we save the scaling factors required for quantization in each layer. Right now the quantization happens in real time and there is no use of optimized kernels for quantized inference.
You can quantize the weights and save them in int8 format to  see the memory reduction, but right now we don't support it. If you would like to do something like that yourself and contribute to the library you are most welcome.

As i said the model doesn't use any optimized kernels for quantized inference and the VNNI instructions are not used.

**Comment by RyanHuangNLP** [comment]
@ofirzaf Thank for response~~so the results is eval on the fp32 one?

**Comment by ofirzaf** [comment]
All the computation is done using tensors with dtype fp32. When training we use the method of fake quantization as described in [Quantization Aware Training](http://nlp_architect.nervanasys.com/quantized_bert.html#quantization-aware-training), implemented here [`training_quantized_forward`](https://github.com/NervanaSystems/nlp-architect/blob/d73e4a4e1dc72e618e36cf2b26ddba3eee3f0c4a/nlp_architect/nn/torch/quantization.py#L113).
During the evaluation of the model we still use tensors with dtype fp32 but the values are int8 numbers in the range of [-127, 127] which should yield the same result as if you were using real int8 tensors. Implemented in here [`inference_quantized_forward`](https://github.com/NervanaSystems/nlp-architect/blob/d73e4a4e1dc72e618e36cf2b26ddba3eee3f0c4a/nlp_architect/nn/torch/quantization.py#L137).

**Comment by RyanHuangNLP** [comment]
@ofirzaf thanks for answer. I think it should be add this detail in the docs. 
And Did you test the inference time between int8 and fp32?

**Comment by ofirzaf** [comment]
No, since we didn't implement optimized kernels for int8 operations we didn't test inference time. Our goal was to show that from the perspective of model accuracy, quantization to 8bit is possible.

**Comment by RyanHuangNLP** [comment]
@ofirzaf Got it~~Thank you very much
